### PR TITLE
[stable/cosbench]: add Cosbench chart

### PR DESCRIPTION
**Pull request by giacope**
#### What this PR does / why we need it:

This PR adds a new chart for installing [COSBench](https://github.com/intel-cloud/cosbench) on Kubernetes.
COSBench is a benchmarking tool to measure the performance of Cloud Object Storage services. It is compatible with OpenStack Swift, Amazon S3, Scality Zenko, Ceph, CDMI, Google Cloud Storage, and other services.

#### Checklist

- [x] [DCO](https://www.helm.sh/blog/helm-dco/index.html) signed
- [x] Variables are documented in the README.md

### CODE REVIEW

File: stable/cosbench/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+name: cosbench
+version: 0.1.0
```
Comment (davidkarlsen) [inline stable/cosbench/Chart.yaml]: set 1.0.0

**Review by davidkarlsen — CHANGES_REQUESTED** [review]

File: stable/cosbench/templates/controller-deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
+apiVersion: extensions/v1beta1
```
Comment (davidkarlsen) [inline stable/cosbench/templates/controller-deployment.yaml]: use api v1

**Review by davidkarlsen — CHANGES_REQUESTED** [review]

File: stable/cosbench/templates/driver-statefulset.yaml [hunk 3]
```diff
@@ -0,0 +1,58 @@
+apiVersion: apps/v1beta1
```
Comment (davidkarlsen) [inline stable/cosbench/templates/driver-statefulset.yaml]: v1

File: stable/cosbench/templates/driver-statefulset.yaml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    release: "{{ .Release.Name }}"
+    heritage: "{{ .Release.Service }}"
+spec:
+  serviceName: {{ template "cosbench-driver.fullname" . }}
+  replicas: {{ .Values.driver.replicaCount }}
+  template:
+    metadata:
+      labels:
+        app: {{ template "cosbench.name" . }}
+        chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
```
Comment (davidkarlsen) [inline stable/cosbench/templates/driver-statefulset.yaml]: afaik labels are immutable in this section (try changing the version locally and do an upgrade and test) - so labels with non-chaning values between upgrades would be needed

File: stable/cosbench/templates/driver-statefulset.yaml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        {{- if eq .Values.driver.antiAffinity "hard" }}
+          requiredDuringSchedulingIgnoredDuringExecution:
+          - topologyKey: "kubernetes.io/hostname"
+            labelSelector:
+              matchLabels:
+                app:  {{ template "cosbench.name" . }}
+                release: {{ .Release.Name | quote }}
+        {{- else if eq .Values.driver.antiAffinity "soft" }}
+          preferredDuringSchedulingIgnoredDuringExecution:
+          - weight: 5
```
Comment (davidkarlsen) [inline stable/cosbench/templates/driver-statefulset.yaml]: should the weight be configurable?

File: stable/cosbench/templates/driver-statefulset.yaml (hunk 4 above)
Comment (ssalaues) [reply stable/cosbench/templates/driver-statefulset.yaml]: Yeah this would definitely cause upgrade issues because of the chart version.

Comment (giacope) [reply stable/cosbench/templates/driver-statefulset.yaml]: Not in the spec.template.metadata.labels, but they are immutable in the `spec.selector.matchLabels`, which has now been added to satisfy `apps/v1`.

File: stable/cosbench/templates/driver-statefulset.yaml (hunk 5 above)
Comment (giacope) [reply stable/cosbench/templates/driver-statefulset.yaml]: Not entirely sure, but I added it just in case.

File: stable/cosbench/Chart.yaml (hunk 1 above)
Comment (giacope) [reply stable/cosbench/Chart.yaml]: Sure I just update it.

Just wondering if this only applies to charts going to stable or is it a new rule (or one I missed)? Afaik, semantic versioning starts at `0.1.0`.

Comment (davidkarlsen) [reply stable/cosbench/Chart.yaml]: When going to stable we start at 1.0.0

File: stable/cosbench/templates/driver-statefulset.yaml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+          requiredDuringSchedulingIgnoredDuringExecution:
+          - topologyKey: "kubernetes.io/hostname"
+            labelSelector:
+              matchLabels:
+                app:  {{ template "cosbench.name" . }}
+                release: {{ .Release.Name | quote }}
+                role: driver
+        {{- else if eq .Values.driver.antiAffinity "soft" }}
+          preferredDuringSchedulingIgnoredDuringExecution:
+          - weight: {{ default 5 .Values.driver.antiAffinityWeight }}
```
Comment (davidkarlsen) [inline stable/cosbench/templates/driver-statefulset.yaml]: we avoid using default and have the defaults in the values.yaml instead

**Review by davidkarlsen — CHANGES_REQUESTED** [review]

File: stable/cosbench/templates/driver-statefulset.yaml (hunk 6 above)
Comment (giacope) [reply stable/cosbench/templates/driver-statefulset.yaml]: Fixed

**Comment by giacomoguiulfo** [comment]
@davidkarlsen Fixed! 👍

File: stable/cosbench/templates/controller-deployment.yaml [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+      release: {{ .Release.Name }}
+      role: controller
+  template:
+    metadata:
+      labels:
+        app: {{ template "cosbench.name" . }}
+        release: {{ .Release.Name }}
+        role: controller
+    spec:
+      containers:
```
Comment (davidkarlsen) [inline stable/cosbench/templates/controller-deployment.yaml:26]: One thing I missed in the previous review was RBAC: see https://github.com/helm/helm/blob/master/docs/chart_best_practices/rbac.md for howto
If there are no interactions with the actual k8s api, just add a service account, if there is interaction with api - add roles and rolebindings too.

**Review by davidkarlsen — CHANGES_REQUESTED** [review]

File: stable/cosbench/templates/controller-deployment.yaml (hunk 7 above)
Comment (giacope) [reply stable/cosbench/templates/controller-deployment.yaml:26]: Good point, I just added `serviceAccounts`.

**Comment by giacomoguiulfo** [comment]
@davidkarlsen Done 👍

**Review by davidkarlsen — APPROVED** [review]

**Pull request merged**